Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems.  All rights reserved.
+############################################################################## */
+IMPORT $.^.Types;
+SliceExt := Types.SliceExt;
+TrainingDat := Types.TrainingDat;
+/**
+  * Neural Network Training for SentenceVectors.
+  * <p>Train specialized SentenceVector neural network given a batch of training
+  * data. Takes in weights as a set of weights slices (SliceExt), and returns
+  * a set of weight adjustments, also formatted as slices.
+  * @param wts The weights slices.
+  * @param train The batch of training data formatted as a main word and set of context words.
+  * @param slicesize The maximum number of weights in a slice.
+  * @param nWeights The total number of weights across all slices.
+  * @param numwords The number of words in the vocabulary.
+  * @param dim The dimensionality of the vectors being trained
+  * @param mbsize The number of training records in the mini-batch.
+  * @param lr The learning rate to use for this batch.
+  * @param negsamp The number of negative samples to choose for each main word.
+  * @return weight updates as DATASET(SliceExt).  Note that these are additive 
+  *         changes to the weights, not the final weight values.
+  */
+EXPORT STREAMED DATASET(SliceExt) svTrainNN(STREAMED DATASET(SliceExt) wts,
+          STREAMED DATASET(trainingDat) train,
+          UNSIGNED4 slicesize,
+          UNSIGNED4 nWeights,
+          UNSIGNED4 numwords,
+          UNSIGNED4 dim,
+          UNSIGNED4 mbsize,
+          REAL lr,
+          UNSIGNED4 negsamp) := EMBED(C++ : activity)
+  #include <stdlib.h>
+  #include <stdio.h>
+  #include <stdint.h>
+  #include <string.h>
+  #include <math.h>
+  #include <assert.h>
+  /**
+  * Training Point Structure
+  */
+  struct tp
+  {
+    uint32_t main;
+    uint32_t contxCount;
+    uint32_t * contx;
+  };
+  /**
+    * Specialized Neural Network with 3 Layers.
+    * The input layer is the size of the vocabulary, the hidden layer is the
+    * size of the word vectors, and the output layer is the size of the vocabulary.
+    * Training data is presented as
+    * a tuple of a main word (id), and a list of context words (ids). Context
+    * is generally the words that were near or in the same sentence as the main
+    * word.
+    * The context words are simultaneously active at the input layer, and the
+    * main word is the training target at the output layer.
+    * The hidden layer weights represent the word vectors.  Since only
```

Comment:
"Since only" is either incomplete or extra

Reply:
removed